examples/udev.rs

```diff
@@ -0,0 +1,477 @@
+use slog::{Drain, Logger};
+use smithay::backend::drm::{DrmBackend, DrmDevice, DrmHandler};
+use smithay::backend::graphics::GraphicsBackend;
+use smithay::backend::graphics::egl::EGLGraphicsBackend;
+use smithay::backend::input::{self, Event, InputBackend, InputHandler, KeyboardKeyEvent, PointerButtonEvent,
+                              PointerAxisEvent};
+use smithay::backend::libinput::{LibinputInputBackend, libinput_bind, PointerAxisEvent as LibinputPointerAxisEvent, LibinputSessionInterface};
+use smithay::backend::udev::{UdevBackend, UdevHandler, udev_backend_bind};
+use smithay::backend::session::{Session, SessionNotifier};
+use smithay::backend::session::direct::{direct_session_bind, DirectSession};
+use smithay::wayland::compositor::{CompositorToken, SubsurfaceRole, TraversalAction};
+use smithay::wayland::compositor::roles::Role;
+use smithay::wayland::output::{Mode, Output, PhysicalProperties};
+use smithay::wayland::seat::{KeyboardHandle, PointerHandle, Seat};
+use smithay::wayland::shell::ShellState;
+use smithay::wayland::shm::init_shm_global;
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::io::Error as IoError;
+use std::rc::Rc;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::Duration;
+use std::process::Command;
+use xkbcommon::xkb::keysyms as xkb;
+use wayland_server::{StateToken, StateProxy};
+use wayland_server::protocol::{wl_output, wl_pointer};
+
+struct LibinputInputHandler {
+    log: Logger,
+    pointer: PointerHandle,
+    keyboard: KeyboardHandle,
+    window_map: Rc<RefCell<MyWindowMap>>,
+    pointer_location: Rc<RefCell<(f64, f64)>>,
+    screen_size: (u32, u32),
+    serial: u32,
+    running: Arc<AtomicBool>,
+}
+
+impl LibinputInputHandler {
+    fn next_serial(&mut self) -> u32 {
+        self.serial += 1;
+        self.serial
+    }
+}
+
+impl InputHandler<LibinputInputBackend> for LibinputInputHandler {
+    fn on_seat_created(&mut self, _: &input::Seat) {
+        /* we just create a single static one */
+    }
+    fn on_seat_destroyed(&mut self, _: &input::Seat) {
+        /* we just create a single static one */
+    }
+    fn on_seat_changed(&mut self, _: &input::Seat) {
+        /* we just create a single static one */
+    }
+    fn on_keyboard_key(&mut self, _: &input::Seat, evt: event::keyboard::KeyboardKeyEvent) {
+        let keycode = evt.key();
+        let state = evt.state();
+        debug!(self.log, "key"; "keycode" => keycode, "state" => format!("{:?}", state));
+
+        let serial = self.next_serial();
+        self.keyboard.input(keycode, state, serial, |modifiers, keysym| {
+            if modifiers.ctrl && modifiers.alt && keysym == xkb::KEY_BackSpace {
+                self.running.store(false, Ordering::SeqCst);
+                false
+            } else if modifiers.logo && keysym == xkb::KEY_Return {
+                let _ = Command::new("weston-terminal").spawn();
+                false
```

Comment:
This should only trigger if `state == Pressed`, otherwise a terminal is spawned on key press *and* on key release, so two instances are spawned each time